include/sas.h

```diff
@@ -304,6 +304,10 @@ class SAS
                                     const char *fmt,
                                     ...);
 
+  typedef int (socket_factory_t)(const char* hostname,
+                                 const char* port);
```

Comment:
One final comment - I would prefer this the be called a callback than a factory (`create_connected_socket_cb` or something). Unlike python, C++ has more of a distinction between functions and objects and I think this terminology will be more familiar to C++ programmers.